src/cdk-experimental/scrolling/virtual-scroll-viewport.ts

```diff
@@ -129,83 +198,84 @@ export class CdkVirtualScrollViewport implements OnInit, OnDestroy {
         this._renderedRangeSubject.next(this._renderedRange = range);
         this._changeDetectorRef.markForCheck();
         this._ngZone.runOutsideAngular(() => this._ngZone.onStable.pipe(take(1)).subscribe(() => {
-          this._scrollStrategy.onContentRendered();
+          // Queue this up in a `Promise.resolve()` so that if the user makes a series of calls
+          // like:
+          //
+          // viewport.setRenderedRange(...);
+          // viewport.setTotalContentSize(...);
+          // viewport.setRenderedContentOffset(...);
+          //
+          // The call to `onContentRendered` will happen after all of the updates have been applied.
+          Promise.resolve().then(() => this._scrollStrategy.onContentRendered());
         }));
       });
     }
   }
 
   /** Sets the offset of the rendered portion of the data from the start (in pixels). */
-  setRenderedContentOffset(offset: number) {
-    const transform =
-        this.orientation === 'horizontal' ? `translateX(${offset}px)` : `translateY(${offset}px)`;
+  setRenderedContentOffset(offset: number, to: 'to-start' | 'to-end' = 'to-start') {
+    const axis = this.orientation === 'horizontal' ? 'X' : 'Y';
+    let transform = `translate${axis}(${Number(offset)}px)`;
+    if (to === 'to-end') {
+      // TODO(mmalerba): The viewport should rewrite this as a `to-start` offset on the next render
+      // cycle. Otherwise elements will appear to expand in the wrong direction (e.g.
+      // `mat-expansion-panel` would expand upward).
+      transform += ` translate${axis}(-100%)`;
+    }
     if (this._renderedContentTransform != transform) {
       // Re-enter the Angular zone so we can mark for change detection.
       this._ngZone.run(() => {
-        this._renderedContentTransform = transform;
+        // We know this value is safe because we parse `offset` with `Number()` before passing it
+        // into the string.
+        this._renderedContentTransform = this._sanitizer.bypassSecurityTrustStyle(transform);
         this._changeDetectorRef.markForCheck();
       });
     }
   }
 
-  /** Attaches a `CdkVirtualForOf` to this viewport. */
-  attach(forOf: CdkVirtualForOf<any>) {
-    if (this._isAttached) {
-      throw Error('CdkVirtualScrollViewport is already attached.');
-    }
-
-    this._isAttached = true;
-    // Subscribe to the data stream of the CdkVirtualForOf to keep track of when the data length
-    // changes.
-    forOf.dataStream.pipe(takeUntil(this._detachedSubject)).subscribe(data => {
-      const len = data.length;
-      if (len != this._dataLength) {
-        this._dataLength = len;
-        this._scrollStrategy.onDataLengthChanged();
-      }
+  /** Sets the scroll offset on the viewport. */
+  setScrollOffset(offset: number) {
+    this._ngZone.run(() => {
+      this._pendingScrollOffset = offset;
```

Comment:
Rather than having this go through change detection, why not set the offset directly on the element?

Reply:
For the sake of performance I'm trying to, as much as possible, batch all DOM writes together